// nc-plugin/build/main/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.XcWebhookNotificationPlugin = exports.XcEmailPlugin = exports.XcStoragePlugin = exports.XcPluginMigration = exports.XcPlugin = exports.XcPluginHooks = void 0;
const XcPlugin_1 = __importDefault(require("./lib/XcPlugin"));
exports.XcPlugin = XcPlugin_1.default;
const XcPluginHooks_1 = __importDefault(require("./lib/XcPluginHooks"));
exports.XcPluginHooks = XcPluginHooks_1.default;
const XcPluginMigration_1 = __importDefault(require("./lib/XcPluginMigration"));
exports.XcPluginMigration = XcPluginMigration_1.default;
const XcStoragePlugin_1 = __importDefault(require("./lib/XcStoragePlugin"));
exports.XcStoragePlugin = XcStoragePlugin_1.default;
const XcEmailPlugin_1 = __importDefault(require("./lib/XcEmailPlugin"));
exports.XcEmailPlugin = XcEmailPlugin_1.default;
const XcWebhookNotificationPlugin_1 = __importDefault(require("./lib/XcWebhookNotificationPlugin"));
exports.XcWebhookNotificationPlugin = XcWebhookNotificationPlugin_1.default;
