// nocodb-sdk/build/main/lib/globals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TiptapMarksTypes = exports.TiptapNodesTypes = exports.ProjectStatus = exports.ModelTypes = exports.PluginCategory = exports.AuditOperationSubTypes = exports.AuditOperationTypes = exports.ErrorMessages = exports.ExportTypes = exports.RelationTypes = exports.ProjectTypes = exports.ViewTypes = void 0;
var ViewTypes;
(function (ViewTypes) {
    ViewTypes[ViewTypes["FORM"] = 1] = "FORM";
    ViewTypes[ViewTypes["GALLERY"] = 2] = "GALLERY";
    ViewTypes[ViewTypes["GRID"] = 3] = "GRID";
    ViewTypes[ViewTypes["KANBAN"] = 4] = "KANBAN";
    ViewTypes[ViewTypes["MAP"] = 5] = "MAP";
})(ViewTypes = exports.ViewTypes || (exports.ViewTypes = {}));
var ProjectTypes;
(function (ProjectTypes) {
    ProjectTypes["DATABASE"] = "database";
    ProjectTypes["DOCUMENTATION"] = "documentation";
    ProjectTypes["COWRITER"] = "cowriter";
    ProjectTypes["DASHBOARD"] = "dashboard";
})(ProjectTypes = exports.ProjectTypes || (exports.ProjectTypes = {}));
var RelationTypes;
(function (RelationTypes) {
    RelationTypes["HAS_MANY"] = "hm";
    RelationTypes["BELONGS_TO"] = "bt";
    RelationTypes["MANY_TO_MANY"] = "mm";
})(RelationTypes = exports.RelationTypes || (exports.RelationTypes = {}));
var ExportTypes;
(function (ExportTypes) {
    ExportTypes["EXCEL"] = "excel";
    ExportTypes["CSV"] = "csv";
})(ExportTypes = exports.ExportTypes || (exports.ExportTypes = {}));
var ErrorMessages;
(function (ErrorMessages) {
    ErrorMessages["INVALID_SHARED_VIEW_PASSWORD"] = "INVALID_SHARED_VIEW_PASSWORD";
    ErrorMessages["NOT_IMPLEMENTED"] = "NOT_IMPLEMENTED";
})(ErrorMessages = exports.ErrorMessages || (exports.ErrorMessages = {}));
var AuditOperationTypes;
(function (AuditOperationTypes) {
    AuditOperationTypes["COMMENT"] = "COMMENT";
    AuditOperationTypes["DATA"] = "DATA";
    AuditOperationTypes["PROJECT"] = "PROJECT";
    AuditOperationTypes["VIRTUAL_RELATION"] = "VIRTUAL_RELATION";
    AuditOperationTypes["RELATION"] = "RELATION";
    AuditOperationTypes["TABLE_VIEW"] = "TABLE_VIEW";
    AuditOperationTypes["TABLE"] = "TABLE";
    AuditOperationTypes["VIEW"] = "VIEW";
    AuditOperationTypes["META"] = "META";
    AuditOperationTypes["WEBHOOKS"] = "WEBHOOKS";
    AuditOperationTypes["AUTHENTICATION"] = "AUTHENTICATION";
    AuditOperationTypes["TABLE_COLUMN"] = "TABLE_COLUMN";
    AuditOperationTypes["ORG_USER"] = "ORG_USER";
})(AuditOperationTypes = exports.AuditOperationTypes || (exports.AuditOperationTypes = {}));
var AuditOperationSubTypes;
(function (AuditOperationSubTypes) {
    AuditOperationSubTypes["UPDATE"] = "UPDATE";
    AuditOperationSubTypes["INSERT"] = "INSERT";
    AuditOperationSubTypes["BULK_INSERT"] = "BULK_INSERT";
    AuditOperationSubTypes["BULK_UPDATE"] = "BULK_UPDATE";
    AuditOperationSubTypes["BULK_DELETE"] = "BULK_DELETE";
    AuditOperationSubTypes["LINK_RECORD"] = "LINK_RECORD";
    AuditOperationSubTypes["UNLINK_RECORD"] = "UNLINK_RECORD";
    AuditOperationSubTypes["DELETE"] = "DELETE";
    AuditOperationSubTypes["CREATE"] = "CREATE";
    AuditOperationSubTypes["RENAME"] = "RENAME";
    AuditOperationSubTypes["IMPORT_FROM_ZIP"] = "IMPORT_FROM_ZIP";
    AuditOperationSubTypes["EXPORT_TO_FS"] = "EXPORT_TO_FS";
    AuditOperationSubTypes["EXPORT_TO_ZIP"] = "EXPORT_TO_ZIP";
    AuditOperationSubTypes["SIGNIN"] = "SIGNIN";
    AuditOperationSubTypes["SIGNUP"] = "SIGNUP";
    AuditOperationSubTypes["PASSWORD_RESET"] = "PASSWORD_RESET";
    AuditOperationSubTypes["PASSWORD_FORGOT"] = "PASSWORD_FORGOT";
    AuditOperationSubTypes["PASSWORD_CHANGE"] = "PASSWORD_CHANGE";
    AuditOperationSubTypes["EMAIL_VERIFICATION"] = "EMAIL_VERIFICATION";
    AuditOperationSubTypes["ROLES_MANAGEMENT"] = "ROLES_MANAGEMENT";
    AuditOperationSubTypes["INVITE"] = "INVITE";
    AuditOperationSubTypes["RESEND_INVITE"] = "RESEND_INVITE";
})(AuditOperationSubTypes = exports.AuditOperationSubTypes || (exports.AuditOperationSubTypes = {}));
var PluginCategory;
(function (PluginCategory) {
    PluginCategory["STORAGE"] = "Storage";
    PluginCategory["EMAIL"] = "Email";
})(PluginCategory = exports.PluginCategory || (exports.PluginCategory = {}));
var ModelTypes;
(function (ModelTypes) {
    ModelTypes["TABLE"] = "table";
    ModelTypes["VIEW"] = "view";
})(ModelTypes = exports.ModelTypes || (exports.ModelTypes = {}));
var ProjectStatus;
(function (ProjectStatus) {
    ProjectStatus["JOB"] = "job";
})(ProjectStatus = exports.ProjectStatus || (exports.ProjectStatus = {}));
var TiptapNodesTypes;
(function (TiptapNodesTypes) {
    TiptapNodesTypes["doc"] = "doc";
    TiptapNodesTypes["sec"] = "sec";
    TiptapNodesTypes["paragraph"] = "paragraph";
    TiptapNodesTypes["text"] = "text";
    TiptapNodesTypes["heading"] = "heading";
    TiptapNodesTypes["bullet"] = "bullet";
    TiptapNodesTypes["ordered"] = "ordered";
    TiptapNodesTypes["task"] = "task";
    TiptapNodesTypes["quote"] = "quote";
    TiptapNodesTypes["divider"] = "divider";
    TiptapNodesTypes["codeBlock"] = "codeBlock";
    TiptapNodesTypes["image"] = "image";
    TiptapNodesTypes["callout"] = "callout";
    TiptapNodesTypes["tipCallout"] = "tipCallout";
    TiptapNodesTypes["table"] = "table";
    TiptapNodesTypes["tableRow"] = "tableRow";
    TiptapNodesTypes["tableCell"] = "tableCell";
    TiptapNodesTypes["embed"] = "embed";
    TiptapNodesTypes["collapsable"] = "collapsable";
    TiptapNodesTypes["collapsableContent"] = "collapsable_content";
    TiptapNodesTypes["collapsableHeader"] = "collapsable_header";
    TiptapNodesTypes["column"] = "column";
    TiptapNodesTypes["columnContent"] = "columnContent";
    TiptapNodesTypes["linkToPage"] = "linkToPage";
    TiptapNodesTypes["attachment"] = "attachment";
})(TiptapNodesTypes = exports.TiptapNodesTypes || (exports.TiptapNodesTypes = {}));
var TiptapMarksTypes;
(function (TiptapMarksTypes) {
    TiptapMarksTypes["strike"] = "strike";
    TiptapMarksTypes["bold"] = "bold";
    TiptapMarksTypes["italic"] = "italic";
    TiptapMarksTypes["link"] = "link";
    TiptapMarksTypes["code"] = "code";
    TiptapMarksTypes["underline"] = "underline";
})(TiptapMarksTypes = exports.TiptapMarksTypes || (exports.TiptapMarksTypes = {}));
