// nocodb-sdk/build/main/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TemplateGenerator = exports.CustomAPI = exports.isLinksOrLTAR = exports.isVirtualCol = exports.isNumericCol = exports.numericUITypes = exports.UITypes = void 0;
__exportStar(require("../lib/XcUIBuilder"), exports);
__exportStar(require("../lib/XcNotification"), exports);
__exportStar(require("../lib/Api"), exports);
__exportStar(require("../lib/columnRules"), exports);
__exportStar(require("../lib/sqlUi"), exports);
__exportStar(require("../lib/globals"), exports);
__exportStar(require("../lib/helperFunctions"), exports);
__exportStar(require("../lib/enums"), exports);
__exportStar(require("../lib/formulaHelpers"), exports);
var UITypes_1 = require("../lib/UITypes");
Object.defineProperty(exports, "UITypes", { enumerable: true, get: function () { return __importDefault(UITypes_1).default; } });
Object.defineProperty(exports, "numericUITypes", { enumerable: true, get: function () { return UITypes_1.numericUITypes; } });
Object.defineProperty(exports, "isNumericCol", { enumerable: true, get: function () { return UITypes_1.isNumericCol; } });
Object.defineProperty(exports, "isVirtualCol", { enumerable: true, get: function () { return UITypes_1.isVirtualCol; } });
Object.defineProperty(exports, "isLinksOrLTAR", { enumerable: true, get: function () { return UITypes_1.isLinksOrLTAR; } });
var CustomAPI_1 = require("../lib/CustomAPI");
Object.defineProperty(exports, "CustomAPI", { enumerable: true, get: function () { return __importDefault(CustomAPI_1).default; } });
var TemplateGenerator_1 = require("../lib/TemplateGenerator");
Object.defineProperty(exports, "TemplateGenerator", { enumerable: true, get: function () { return __importDefault(TemplateGenerator_1).default; } });
__exportStar(require("../lib/passwordHelpers"), exports);
__exportStar(require("../lib/mergeSwaggerSchema"), exports);
