// xc-core-ts/dist/lib/xc-core-ts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ApiInterface_1 = require("./lib/ApiInterface");
Object.defineProperty(exports, "FindClause", { enumerable: true, get: function () { return ApiInterface_1.FindClause; } });
var BaseComponent_1 = require("./lib/BaseComponent");
Object.defineProperty(exports, "BaseComponent", { enumerable: true, get: function () { return BaseComponent_1.BaseComponent; } });
var BaseService_1 = require("./lib/BaseService");
Object.defineProperty(exports, "BaseService", { enumerable: true, get: function () { return BaseService_1.BaseService; } });
var BaseRouter_1 = require("./lib/BaseRouter");
Object.defineProperty(exports, "BaseRouter", { enumerable: true, get: function () { return BaseRouter_1.BaseRouter; } });
var BaseResolver_1 = require("./lib/BaseResolver");
Object.defineProperty(exports, "BaseResolver", { enumerable: true, get: function () { return BaseResolver_1.BaseResolver; } });
var BaseType_1 = require("./lib/BaseType");
Object.defineProperty(exports, "BaseType", { enumerable: true, get: function () { return BaseType_1.BaseType; } });
var CoreRest_1 = require("./lib/CoreRest");
Object.defineProperty(exports, "RouteType", { enumerable: true, get: function () { return CoreRest_1.RouteType; } });
var Loader_1 = require("./lib/Loader");
Object.defineProperty(exports, "Loader", { enumerable: true, get: function () { return Loader_1.Loader; } });
var BaseMiddleware_1 = require("./lib/BaseMiddleware");
Object.defineProperty(exports, "BaseMiddleware", { enumerable: true, get: function () { return BaseMiddleware_1.BaseMiddleware; } });
var BaseMiddlewareGql_1 = require("./lib/BaseMiddlewareGql");
Object.defineProperty(exports, "BaseMiddlewareGql", { enumerable: true, get: function () { return BaseMiddlewareGql_1.BaseMiddlewareGql; } });
