// nocodb-sdk/build/module/lib/formulaHelpers.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import jsep from 'jsep';
export const jsepCurlyHook = {
    name: 'curly',
    init(jsep) {
        jsep.hooks.add('gobble-token', function gobbleCurlyLiteral(env) {
            const OCURLY_CODE = 123; // {
            const CCURLY_CODE = 125; // }
            let start = -1;
            const { context } = env;
            if (!jsep.isIdentifierStart(context.code) &&
                context.code === OCURLY_CODE) {
                if (start == -1) {
                    start = context.index;
                }
                context.index += 1;
                context.gobbleExpressions(CCURLY_CODE);
                if (context.code === CCURLY_CODE) {
                    context.index += 1;
                    env.node = {
                        type: jsep.IDENTIFIER,
                        name: /{{(.*?)}}/.test(context.expr)
                            ? // start would be the position of the first curly bracket
                                // add 2 to point to the first character for expressions like {{col1}}
                                context.expr.slice(start + 2, context.index - 1)
                            : // start would be the position of the first curly bracket
                                // add 1 to point to the first character for expressions like {col1}
                                context.expr.slice(start + 1, context.index - 1),
                    };
                    return env.node;
                }
                else {
                    context.throwError('Unclosed }');
                }
            }
        });
    },
};
export function substituteColumnAliasWithIdInFormula(formula, columns) {
    return __awaiter(this, void 0, void 0, function* () {
        const substituteId = (pt) => __awaiter(this, void 0, void 0, function* () {
            if (pt.type === 'CallExpression') {
                for (const arg of pt.arguments || []) {
                    yield substituteId(arg);
                }
            }
            else if (pt.type === 'Literal') {
                return;
            }
            else if (pt.type === 'Identifier') {
                const colNameOrId = pt.name;
                const column = columns.find((c) => c.id === colNameOrId ||
                    c.column_name === colNameOrId ||
                    c.title === colNameOrId);
                pt.name = '{' + column.id + '}';
            }
            else if (pt.type === 'BinaryExpression') {
                yield substituteId(pt.left);
                yield substituteId(pt.right);
            }
        });
        // register jsep curly hook
        jsep.plugins.register(jsepCurlyHook);
        const parsedFormula = jsep(formula);
        yield substituteId(parsedFormula);
        return jsepTreeToFormula(parsedFormula);
    });
}
export function substituteColumnIdWithAliasInFormula(formula, columns, rawFormula) {
    const substituteId = (pt, ptRaw) => {
        var _a;
        if (pt.type === 'CallExpression') {
            let i = 0;
            for (const arg of pt.arguments || []) {
                substituteId(arg, (_a = ptRaw === null || ptRaw === void 0 ? void 0 : ptRaw.arguments) === null || _a === void 0 ? void 0 : _a[i++]);
            }
        }
        else if (pt.type === 'Literal') {
            return;
        }
        else if (pt.type === 'Identifier') {
            const colNameOrId = pt === null || pt === void 0 ? void 0 : pt.name;
            const column = columns.find((c) => c.id === colNameOrId ||
                c.column_name === colNameOrId ||
                c.title === colNameOrId);
            pt.name = (column === null || column === void 0 ? void 0 : column.title) || (ptRaw === null || ptRaw === void 0 ? void 0 : ptRaw.name) || (pt === null || pt === void 0 ? void 0 : pt.name);
        }
        else if (pt.type === 'BinaryExpression') {
            substituteId(pt.left, ptRaw === null || ptRaw === void 0 ? void 0 : ptRaw.left);
            substituteId(pt.right, ptRaw === null || ptRaw === void 0 ? void 0 : ptRaw.right);
        }
    };
    // register jsep curly hook
    jsep.plugins.register(jsepCurlyHook);
    const parsedFormula = jsep(formula);
    const parsedRawFormula = rawFormula && jsep(rawFormula);
    substituteId(parsedFormula, parsedRawFormula);
    return jsepTreeToFormula(parsedFormula);
}
export function jsepTreeToFormula(node) {
    if (node.type === 'BinaryExpression' || node.type === 'LogicalExpression') {
        return ('(' +
            jsepTreeToFormula(node.left) +
            ' ' +
            node.operator +
            ' ' +
            jsepTreeToFormula(node.right) +
            ')');
    }
    if (node.type === 'UnaryExpression') {
        return node.operator + jsepTreeToFormula(node.argument);
    }
    if (node.type === 'MemberExpression') {
        return (jsepTreeToFormula(node.object) +
            '[' +
            jsepTreeToFormula(node.property) +
            ']');
    }
    if (node.type === 'Identifier') {
        const formulas = [
            'AVG',
            'ADD',
            'DATEADD',
            'DATETIME_DIFF',
            'WEEKDAY',
            'AND',
            'OR',
            'CONCAT',
            'TRIM',
            'UPPER',
            'LOWER',
            'LEN',
            'MIN',
            'MAX',
            'CEILING',
            'FLOOR',
            'ROUND',
            'MOD',
            'REPEAT',
            'LOG',
            'EXP',
            'POWER',
            'SQRT',
            'SQRT',
            'ABS',
            'NOW',
            'REPLACE',
            'SEARCH',
            'INT',
            'RIGHT',
            'LEFT',
            'SUBSTR',
            'MID',
            'IF',
            'SWITCH',
            'URL',
        ];
        if (!formulas.includes(node.name.toUpperCase()))
            return '{' + node.name + '}';
        return node.name;
    }
    if (node.type === 'Literal') {
        if (typeof node.value === 'string') {
            return String.raw `"${escapeLiteral(node.value)}"`;
        }
        return '' + node.value;
    }
    if (node.type === 'CallExpression') {
        return (jsepTreeToFormula(node.callee) +
            '(' +
            node.arguments.map(jsepTreeToFormula).join(', ') +
            ')');
    }
    if (node.type === 'ArrayExpression') {
        return '[' + node.elements.map(jsepTreeToFormula).join(', ') + ']';
    }
    if (node.type === 'Compound') {
        return node.body.map((e) => jsepTreeToFormula(e)).join(' ');
    }
    if (node.type === 'ConditionalExpression') {
        return (jsepTreeToFormula(node.test) +
            ' ? ' +
            jsepTreeToFormula(node.consequent) +
            ' : ' +
            jsepTreeToFormula(node.alternate));
    }
    return '';
}
function escapeLiteral(v) {
    return (v
        // replace \ to \\
        .replace(/\\/g, `\\\\`)
        // replace " to \"
        .replace(/"/g, `\\"`)
        // replace ' to \'
        .replace(/'/g, `\\'`));
}
