// @aws-sdk/client-kafka/dist-cjs/Kafka.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Kafka = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const BatchAssociateScramSecretCommand_1 = require("./commands/BatchAssociateScramSecretCommand");
const BatchDisassociateScramSecretCommand_1 = require("./commands/BatchDisassociateScramSecretCommand");
const CreateClusterCommand_1 = require("./commands/CreateClusterCommand");
const CreateClusterV2Command_1 = require("./commands/CreateClusterV2Command");
const CreateConfigurationCommand_1 = require("./commands/CreateConfigurationCommand");
const CreateVpcConnectionCommand_1 = require("./commands/CreateVpcConnectionCommand");
const DeleteClusterCommand_1 = require("./commands/DeleteClusterCommand");
const DeleteClusterPolicyCommand_1 = require("./commands/DeleteClusterPolicyCommand");
const DeleteConfigurationCommand_1 = require("./commands/DeleteConfigurationCommand");
const DeleteVpcConnectionCommand_1 = require("./commands/DeleteVpcConnectionCommand");
const DescribeClusterCommand_1 = require("./commands/DescribeClusterCommand");
const DescribeClusterOperationCommand_1 = require("./commands/DescribeClusterOperationCommand");
const DescribeClusterOperationV2Command_1 = require("./commands/DescribeClusterOperationV2Command");
const DescribeClusterV2Command_1 = require("./commands/DescribeClusterV2Command");
const DescribeConfigurationCommand_1 = require("./commands/DescribeConfigurationCommand");
const DescribeConfigurationRevisionCommand_1 = require("./commands/DescribeConfigurationRevisionCommand");
const DescribeVpcConnectionCommand_1 = require("./commands/DescribeVpcConnectionCommand");
const GetBootstrapBrokersCommand_1 = require("./commands/GetBootstrapBrokersCommand");
const GetClusterPolicyCommand_1 = require("./commands/GetClusterPolicyCommand");
const GetCompatibleKafkaVersionsCommand_1 = require("./commands/GetCompatibleKafkaVersionsCommand");
const ListClientVpcConnectionsCommand_1 = require("./commands/ListClientVpcConnectionsCommand");
const ListClusterOperationsCommand_1 = require("./commands/ListClusterOperationsCommand");
const ListClusterOperationsV2Command_1 = require("./commands/ListClusterOperationsV2Command");
const ListClustersCommand_1 = require("./commands/ListClustersCommand");
const ListClustersV2Command_1 = require("./commands/ListClustersV2Command");
const ListConfigurationRevisionsCommand_1 = require("./commands/ListConfigurationRevisionsCommand");
const ListConfigurationsCommand_1 = require("./commands/ListConfigurationsCommand");
const ListKafkaVersionsCommand_1 = require("./commands/ListKafkaVersionsCommand");
const ListNodesCommand_1 = require("./commands/ListNodesCommand");
const ListScramSecretsCommand_1 = require("./commands/ListScramSecretsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const ListVpcConnectionsCommand_1 = require("./commands/ListVpcConnectionsCommand");
const PutClusterPolicyCommand_1 = require("./commands/PutClusterPolicyCommand");
const RebootBrokerCommand_1 = require("./commands/RebootBrokerCommand");
const RejectClientVpcConnectionCommand_1 = require("./commands/RejectClientVpcConnectionCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const UpdateBrokerCountCommand_1 = require("./commands/UpdateBrokerCountCommand");
const UpdateBrokerStorageCommand_1 = require("./commands/UpdateBrokerStorageCommand");
const UpdateBrokerTypeCommand_1 = require("./commands/UpdateBrokerTypeCommand");
const UpdateClusterConfigurationCommand_1 = require("./commands/UpdateClusterConfigurationCommand");
const UpdateClusterKafkaVersionCommand_1 = require("./commands/UpdateClusterKafkaVersionCommand");
const UpdateConfigurationCommand_1 = require("./commands/UpdateConfigurationCommand");
const UpdateConnectivityCommand_1 = require("./commands/UpdateConnectivityCommand");
const UpdateMonitoringCommand_1 = require("./commands/UpdateMonitoringCommand");
const UpdateSecurityCommand_1 = require("./commands/UpdateSecurityCommand");
const UpdateStorageCommand_1 = require("./commands/UpdateStorageCommand");
const KafkaClient_1 = require("./KafkaClient");
const commands = {
    BatchAssociateScramSecretCommand: BatchAssociateScramSecretCommand_1.BatchAssociateScramSecretCommand,
    BatchDisassociateScramSecretCommand: BatchDisassociateScramSecretCommand_1.BatchDisassociateScramSecretCommand,
    CreateClusterCommand: CreateClusterCommand_1.CreateClusterCommand,
    CreateClusterV2Command: CreateClusterV2Command_1.CreateClusterV2Command,
    CreateConfigurationCommand: CreateConfigurationCommand_1.CreateConfigurationCommand,
    CreateVpcConnectionCommand: CreateVpcConnectionCommand_1.CreateVpcConnectionCommand,
    DeleteClusterCommand: DeleteClusterCommand_1.DeleteClusterCommand,
    DeleteClusterPolicyCommand: DeleteClusterPolicyCommand_1.DeleteClusterPolicyCommand,
    DeleteConfigurationCommand: DeleteConfigurationCommand_1.DeleteConfigurationCommand,
    DeleteVpcConnectionCommand: DeleteVpcConnectionCommand_1.DeleteVpcConnectionCommand,
    DescribeClusterCommand: DescribeClusterCommand_1.DescribeClusterCommand,
    DescribeClusterOperationCommand: DescribeClusterOperationCommand_1.DescribeClusterOperationCommand,
    DescribeClusterOperationV2Command: DescribeClusterOperationV2Command_1.DescribeClusterOperationV2Command,
    DescribeClusterV2Command: DescribeClusterV2Command_1.DescribeClusterV2Command,
    DescribeConfigurationCommand: DescribeConfigurationCommand_1.DescribeConfigurationCommand,
    DescribeConfigurationRevisionCommand: DescribeConfigurationRevisionCommand_1.DescribeConfigurationRevisionCommand,
    DescribeVpcConnectionCommand: DescribeVpcConnectionCommand_1.DescribeVpcConnectionCommand,
    GetBootstrapBrokersCommand: GetBootstrapBrokersCommand_1.GetBootstrapBrokersCommand,
    GetClusterPolicyCommand: GetClusterPolicyCommand_1.GetClusterPolicyCommand,
    GetCompatibleKafkaVersionsCommand: GetCompatibleKafkaVersionsCommand_1.GetCompatibleKafkaVersionsCommand,
    ListClientVpcConnectionsCommand: ListClientVpcConnectionsCommand_1.ListClientVpcConnectionsCommand,
    ListClusterOperationsCommand: ListClusterOperationsCommand_1.ListClusterOperationsCommand,
    ListClusterOperationsV2Command: ListClusterOperationsV2Command_1.ListClusterOperationsV2Command,
    ListClustersCommand: ListClustersCommand_1.ListClustersCommand,
    ListClustersV2Command: ListClustersV2Command_1.ListClustersV2Command,
    ListConfigurationRevisionsCommand: ListConfigurationRevisionsCommand_1.ListConfigurationRevisionsCommand,
    ListConfigurationsCommand: ListConfigurationsCommand_1.ListConfigurationsCommand,
    ListKafkaVersionsCommand: ListKafkaVersionsCommand_1.ListKafkaVersionsCommand,
    ListNodesCommand: ListNodesCommand_1.ListNodesCommand,
    ListScramSecretsCommand: ListScramSecretsCommand_1.ListScramSecretsCommand,
    ListTagsForResourceCommand: ListTagsForResourceCommand_1.ListTagsForResourceCommand,
    ListVpcConnectionsCommand: ListVpcConnectionsCommand_1.ListVpcConnectionsCommand,
    PutClusterPolicyCommand: PutClusterPolicyCommand_1.PutClusterPolicyCommand,
    RebootBrokerCommand: RebootBrokerCommand_1.RebootBrokerCommand,
    RejectClientVpcConnectionCommand: RejectClientVpcConnectionCommand_1.RejectClientVpcConnectionCommand,
    TagResourceCommand: TagResourceCommand_1.TagResourceCommand,
    UntagResourceCommand: UntagResourceCommand_1.UntagResourceCommand,
    UpdateBrokerCountCommand: UpdateBrokerCountCommand_1.UpdateBrokerCountCommand,
    UpdateBrokerStorageCommand: UpdateBrokerStorageCommand_1.UpdateBrokerStorageCommand,
    UpdateBrokerTypeCommand: UpdateBrokerTypeCommand_1.UpdateBrokerTypeCommand,
    UpdateClusterConfigurationCommand: UpdateClusterConfigurationCommand_1.UpdateClusterConfigurationCommand,
    UpdateClusterKafkaVersionCommand: UpdateClusterKafkaVersionCommand_1.UpdateClusterKafkaVersionCommand,
    UpdateConfigurationCommand: UpdateConfigurationCommand_1.UpdateConfigurationCommand,
    UpdateConnectivityCommand: UpdateConnectivityCommand_1.UpdateConnectivityCommand,
    UpdateMonitoringCommand: UpdateMonitoringCommand_1.UpdateMonitoringCommand,
    UpdateSecurityCommand: UpdateSecurityCommand_1.UpdateSecurityCommand,
    UpdateStorageCommand: UpdateStorageCommand_1.UpdateStorageCommand,
};
class Kafka extends KafkaClient_1.KafkaClient {
}
exports.Kafka = Kafka;
(0, smithy_client_1.createAggregatedClient)(commands, Kafka);
